### PR TITLE
Feature/local attachments

### LINKED ISSUE #251: Possibility to attach local files [$50]
(Carrying over the tasks from https://github.com/nextcloud/mail/pull/4 so we don’t forget it ;)

client-side TODO
- [ ] upload files asynchronously
- [ ] add uploaded files to list of attachments
- [ ] show loading and/or progress indicator
- [ ] lock 'Send'/'Reply' button until all local attachments have been uploaded successfully
- [ ] handle errors (remove faulty attachment from list)
- [ ] add special property to attachment so that server can distinguish between local and cloud attachments 

server-side TODO
- [ ] add attachment store for asynchronously uploaded files
- [ ] move uploaded attachments to temp dir (use core's temp manager)
- [ ] limit uploaded file size to a sane size (e.g. a few MBs)
- [ ] distinguish between local and cloud attachments when sending a message and load local attachments from the attachment store

cc @nextcloud/mail 

<bountysource-plugin>

---
There is a **[$50 open bounty](https://www.bountysource.com/issues/40166193-possibility-to-attach-local-files?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github)** on this issue. Add to the bounty at [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F44154351&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by ChristophWurst**
Updated and cleanup-up version of https://github.com/nextcloud/mail/pull/383, pushed to the main repo.

See the other PR for a description and the original discussion.

Fixes #251

TODO:
- [x] client-side impl by @lcalaresu 
- server-side impl
  - [x] use app data
  - [x] store uploaded file
  - [x] ~~send DELETE request to the server to clean-up attachments that were removed on the client-side~~ -> can be handled by garbage collection as noted below.
  - [x] attach uploaded file when sending a message
  - [x] test code

As a future enhancement we should add some kind of attachment garbage collector that purges the attachment storage for anything that is older than let's say one day.

### CODE REVIEW

**Comment by ChristophWurst** [comment]
@lcalaresu just wanted to thank you again for you hard work on this. The UI part is awesome! I really like the simplicity. The progress bar looks good and gives a nice feedback about the upload state 🚀

File: js/tests/views/attachment_spec.js [hunk 1]
```diff
@@ -1,124 +0,0 @@
...
- */
-
-
-define(['views/attachment',
-		'models/localattachment',
-		'models/attachment',
-		'radio'],
-	function(AttachmentView, LocalAttachment, Attachment, Radio) {
-
-	describe('AttachmentView', function () {
```
Comment (lcalaresu) [inline js/tests/views/attachment_spec.js]: The test has been deleted. You do not want to test the view?

**Comment by lcalaresu** [comment]
Thank you to take it over! Sorry I cannot do mush more than reviewing these days -_-
The PHP part seems easy for you :smile:

File: js/tests/views/attachment_spec.js (hunk 1 above)
Comment (ChristophWurst) [inline js/tests/views/attachment_spec.js]: Whoops. That wasn't my intention. Will revert this change, good catch! :)

Comment (ChristophWurst) [inline js/tests/views/attachment_spec.js]: Has it really been deleted? I don't see that change it the full diff and also GitHub lists your review comment as *outdated*.

**Comment by ChristophWurst** [comment]
> Thank you to take it over! Sorry I cannot do mush more than reviewing these days -_-
> The PHP part seems easy for you 😄

Sure, no problem! Reviewing PRs is as important as developing them and thus very much appreciated that you're so actively involved here! 😄

File: js/tests/views/attachment_spec.js (hunk 1 above)
Comment (lcalaresu) [inline js/tests/views/attachment_spec.js]: Yeah, I saw the outdated part, but can't really see it on the full diff...
It the `js/tests/views/attachment_spec.js` without an `S` on attachment ^_^

Comment (ChristophWurst) [inline js/tests/views/attachment_spec.js]: Found and restored it :)

**Comment by ChristophWurst** [comment]
I ran into a little problem: the server-side code fore sending a message is a huge mess. It's highly complex and error-prone. I tried to add the local attachments logic to it, but then it gets *really* messy.

Since sending a message is a key feature of this app, I will try to implement this enhancement as clean as possible. Hence a little cleanup/refactoring is required, which I'll look into now. This PR is put on hold for now.

**Comment by ChristophWurst** [comment]
> Since sending a message is a key feature of this app, I will try to implement this enhancement as clean as possible. Hence a little cleanup/refactoring is required, which I'll look into now. This PR is put on hold for now.

Cleanup PR can be found at https://github.com/nextcloud/mail/issues/411.

**Comment by ChristophWurst** [comment]
Seems to work now. @nextcloud/mail please test and review :)

**Comment by lcalaresu** [comment]
I just gave it a try, but didn't check the code yet.

However, when I upload an attachment, I get the following error PHP error:
```
"Exception":"DoctrineDBALExceptionNotNullConstraintViolationException",
"Message":"An exception occurred while executing 'INSERT INTO "oc_mail_attachments"("user_id","file_name") VALUES(?,?)' with params ["user", "file.png"]:
   SQLSTATE[23000]: Integrity constraint violation: 19 NOT NULL constraint failed: oc_mail_attachments.file_path",
"Code":0,
"Trace":"#0 /home/dev/github/core/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/DBALException.php(128): DoctrineDBALDriverAbstractSQLiteDriver->convertException('An exception oc...', Object(DoctrineDBALDriverPDOException))
 #1 /home/dev/github/core/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Statement.php(177): DoctrineDBALDBALException::driverExceptionDuringQuery(Object(DoctrineDBALDriverPDOSqliteDriver), Object(DoctrineDBALDriverPDOException), 'INSERT INTO "oc...', Array)
 #2  /home/dev/github/core/lib/public/AppFramework/Db/Mapper.php(247): DoctrineDBALStatement->execute()
 #3  /home/dev/github/core/lib/public/AppFramework/Db/Mapper.php(125): OCPAppFrameworkDbMapper->execute('INSERT INTO `*P...', Array)
 #4  /home/dev/github/core/apps/mail/lib/Service/Attachment/AttachmentService.php(61): OCPAppFrameworkDbMapper->insert(Object(OCAMailDbLocalAttachment))
 #5  /home/dev/github/core/apps/mail/lib/Controller/LocalAttachmentsController.php(66): OCAMailServiceAttachmentAttachmentService->addFile('user', Object(OCAMailServiceAttachmentUploadedFile))
 #6  [internal function]: OCAMailControllerLocalAttachmentsController->create()
 #7  /home/dev/github/core/lib/private/AppFramework/Http/Dispatcher.php(160): call_user_func_array(Array, Array)
 #8  /home/dev/github/core/lib/private/AppFramework/Http/Dispatcher.php(90): OCAppFrameworkHttpDispatcher->executeController(Object(OCAMailControllerLocalAttachmentsController), 'create')
 #9  /home/dev/github/core/lib/private/AppFramework/App.php(114): OCAppFrameworkHttpDispatcher->dispatch(Object(OCAMailControllerLocalAttachmentsController), 'create')
 #10 /home/dev/github/core/lib/private/AppFramework/Routing/RouteActionHandler.php(47): OCAppFrameworkApp::main('OCAMailContro...', 'create', Object(OCAppFrameworkDependencyInjectionDIContainer), Array)
 #11 [internal function]: OCAppFrameworkRoutingRouteActionHandler->__invoke(Array)
 #12 /home/dev/github/core/lib/private/Route/Router.php(299): call_user_func(Object(OCAppFrameworkRoutingRouteActionHandler), Array)
 #13 /home/dev/github/core/lib/base.php(1003): OCRouteRouter->match('/apps/mail/atta...')
 #14 /home/dev/github/core/index.php(40): OC::handleRequest()
 #15 {main}",
"File":"/home/dev/github/core/3rdparty/doctrine/dbal/lib/Doctrine/DBAL/Driver/AbstractSQLiteDriver.php","Line":54
```

I am back from vacation so... maybe I did something wrong though, coz it looks weird...

File: js/views/composerview.js [hunk 2]
```diff
@@ -394,7 +404,7 @@ define(function(require) {
 					onSuccess();
 				}
 
-				if (this.draftUID !== null) {
+				if (_this.draftUID !== null) {
 					// update UID in message list
 					var collection = Radio.ui.request('messagesview:collection');
 					var message = collection.findWhere({id: this.draftUID});
```
Comment (lcalaresu) [inline js/views/composerview.js]: I think `_this` should be used instead... Maybe in a different PR?

**Comment by ChristophWurst** [comment]
> I am back from vacation so... maybe I did something wrong though, coz it looks weird...

I've slightly changed the DB schema. Set the `installed_version` of mail to a lower value so that another migration is run and the extra column is dropped. Let me know if you need help with this :)

File: lib/Service/MailTransmission.php [hunk 3]
```diff
@@ -35,36 +37,42 @@
...
+	private $logger;
+
 	/**
 	 * @param AddressCollector $addressCollector
 	 * @param Folder $userFolder
 	 * @param Logger $logger
 	 */
-	public function __construct(AddressCollector $addressCollector, $userFolder, Logger $logger) {
+	public function __construct(AddressCollector $addressCollector, $userFolder,
+		IAttachmentService $attachmentService, Logger $logger) {
```
Comment (lcalaresu) [inline lib/Service/MailTransmission.php]: attachmentService does not have its associated `@param` entry

File: lib/Service/MailTransmission.php [hunk 4]
```diff
@@ -94,7 +102,8 @@ public function sendMessage(NewMessageData $messageData, RepliedMessageData $rep
 	 * @param RepliedMessageData $replyData
 	 * @return IMessage
 	 */
-	private function buildReplyMessage(Account $account, NewMessageData $messageData, RepliedMessageData $replyData) {
+	private function buildReplyMessage(Account $account,
+		NewMessageData $messageData, RepliedMessageData $replyData) {
```
Comment (lcalaresu) [inline lib/Service/MailTransmission.php]: messageData does not have its associated `@param` entry

**Comment by lcalaresu** [comment]
> I've slightly changed the DB schema. Set the installed_version of mail to a lower value so that another migration is run and the extra column is dropped. Let me know if you need help with this :)

hahaha yeah... Thanks for that :sweat_smile:
It now works much better ^^

Really good job!! :+1::+1:
I tried sending several mails with uploaded attachments and/or from Files, it works as expected.

About file size:
When I upload big files, no problem if my server is correcly configurated. 
However, I get a `Sending mail failed: Message size exceeds fixed limit` PHP error on server with a generic message `Server Error` on my browser.
This is not related to this PR though, because the same problem happens when I send "Attachment from Files".
This PR is already huge, maybe it would be wiser to correct that separately?

File: js/views/composerview.js (hunk 2 above)
Comment (ChristophWurst) [inline js/views/composerview.js]: It doesn't really matter here. `_this` is only really necessary in the closures below

File: lib/Service/MailTransmission.php (hunk 3 above)
Comment (ChristophWurst) [inline lib/Service/MailTransmission.php]: fixed

File: lib/Service/MailTransmission.php (hunk 4 above)
Comment (ChristophWurst) [inline lib/Service/MailTransmission.php]: fixed

**Comment by ChristophWurst** [comment]
> This PR is already huge, maybe it would be wiser to correct that separately?

Seconded. Let's improve this feature step by step. The basics are covered here, any further enhancements should be done in separate PRs :)

**Comment by ChristophWurst** [comment]
We can also add integration tests for this feature similar to what I've started in https://github.com/nextcloud/mail/pull/419

File: js/views/composerview.js (hunk 2 above)
Comment (lcalaresu) [inline js/views/composerview.js]: Actually, I continuously get a JS error while composing my emails:

> Uncaught (in promise) TypeError: Cannot read property 'draftUID' of undefined

But it does not matter here anyway, I think I already talked about that one in another issue about drafts...

**Comment by lcalaresu** [comment]
> Seconded. Let's improve this feature step by step. The basics are covered here, any further enhancements should be done in separate PRs :)

I reckon, This PR is then perfect to me!! Did not encounter any problem while sending maybe two dozens of email.

**Comment by ChristophWurst** [comment]
> Good job 👏
> That was a long awaited feature for me 😄

Awesome. Thanks a lot for your help with the implementation! Now please claim your bounty at https://www.bountysource.com/issues/40166193-possibility-to-attach-local-files 😉

File: js/views/composerview.js (hunk 2 above)
Comment (ChristophWurst) [inline js/views/composerview.js]: Right. When I checked the code I was actually looking at the code responsible for sending new messages, hence I didn't see the issue. Anyway, fixed it in https://github.com/nextcloud/mail/pull/420 :)

**Comment by lcalaresu** [comment]
@ChristophWurst  And thank you for finishing it :wink:

**Comment by jancborchardt** [comment]
Awesome work @lcalaresu! :) Well deserved bounty, and cool you are part of the Mail app team!
